Find recurrent focal CN calls

cbethell opened the pull request:
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section

The purpose of this PR is to find the _recurrent_ focal CN calls using the output of PR #644 .

#### What scientific question is your analysis addressing?

This PR is working to isolate and save the _recurrent_ most focal CN calls.

#### What was your approach?

Using the `consensus_seg_most_focal_cn_status.tsv.gz` output file from PR #644, the focal CN calls were tallied by region (counting how many samples has the same call for the same region).

The final data frame was then created by keeping observations that have more than 3 samples with the same call in the same region.

#### What GitHub issue does your pull request address?

Related: #486

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?

- As mentioned above, I defined the cutoff for recurrence in this PR to be > 3. In the notebook, I provide further evidence for this decision, including a plot of the distribution of counts which should probably receive a particularly close look.
- Is the final file in an ideal format? The current format being: `status`, `region`, `region_type`, `count` -- we may/may not want to remove the count column

#### Is there anything that you want to discuss further?

- I started to cross check the cytoband regions in the final recurrent data frame with the cytoband region calls in the GISTIC results but did not want to go further down that rabbit hole unless that is something we would want at this stage.

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes. Note however that the `README.md` file will be updated once PR #644 is merged to avoid conflicts.

### Results

#### What types of results are included (e.g., table, figure)?

The output of this PR includes a TSV file with the _recurrent_ focal CN calls by region found in the module's directory at:
`results/consensus_seg_recurrent_focal_cn_units.tsv`

The output html notebook can be seen [here](https://cbethell.github.io/open-pbta-output/06-find-recurrent-calls.nb.html)

#### What is your summary of the results?

The final results includes the most focal CN calls made in PR #644 for regions that have more than 3 samples in agreement.

#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [ ] This analysis module has a `README` and it is up to date. Will update once #644 is merged.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

<!-- IF YOUR PULL REQUEST IS A DATA RELEASE, PLEASE REMOVE THE [HTML COMMENT TAG](https://html.com/tags/comment-tag/) FROM THE SECTION BELOW AND COMPLETE THE CHECKLIST-->

<!-- 
### Data Release Checklist
- [ ] Is the table in doc/data-file-descriptions.md up to date?
- [ ] Is doc/data-format.md up to date?
- [ ] Is doc/release-notes.md up to date?
- [ ] Is download-data.sh up to date?
-->

cansavvy commented on analyses/focal-cn-file-preparation/06-find-recurrent-calls.Rmd:
```diff
@@ -0,0 +1,95 @@
+
+This notebook determines the recurrent focal copy number dominant status calls by region using the output of `05-define-most-focal-cn-units.Rmd`.
+
+Recurrence here has been arbitrarily defined based on the plotting of the distribution of status calls and a [similar decision](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/d197f2717eb0d067fe35644db91c4910b6d# 46c06/analyses/fusion_filtering/06-recurrent-fusions-per-histology.R#L111) made in `analyses/fusion_filtering/06-recurrent-fusions-per-histology.R`
```
This link isn't working for me.

cansavvy replied:
Is the similar decision you are referring to here the distribution plot or something else? If it's something else can you describe this here?

cansavvy replied:
Can you add a bit more rationale about how the cutoff has come to be? You may also want to add a a thing in the set up where we can easily change the cutoff if needed.

cbethell replied:
The similar decision made here is the cutoff. I can fix the link and add a better description here.

cbethell replied:
I updated this description in the last commit and in the rendered notebook [here](https://cbethell.github.io/open-pbta-output/06-find-recurrent-calls.nb.html). Let me know if this is along the lines of what you were looking for!

cansavvy commented on analyses/focal-cn-file-preparation/06-find-recurrent-calls.Rmd:
```diff
@@ -0,0 +1,95 @@
+# Tally the number of samples that have the same status call by region
+final_df_tally <- focal_cn_df %>%
+  group_by(status, region, region_type) %>%
+  tally() %>%
```
```suggestion
  tally(name = "count") 
```
Then you can get rid of the renaming step in the next line.

cansavvy commented on analyses/focal-cn-file-preparation/06-find-recurrent-calls.Rmd:
```diff
@@ -0,0 +1,95 @@
+# Plot the distribution of counts
+ggplot(final_df_tally, aes(x = count)) +
+  geom_histogram(bins = 100) +
+  geom_vline(aes(xintercept = 3), linetype = "longdash", color = "red", size = .4)
```
May want to change the color here so it's easier to see the line.

cansavvy commented on analyses/focal-cn-file-preparation/06-find-recurrent-calls.Rmd:
```diff
@@ -0,0 +1,95 @@
+# Here, we arbitrarily define recurrence to be greater than 3 based on 
+# the plot above and on a similar decision made in
+# `analyses/fusion_filtering/06-recurrent-fusions-per-histology.R`
+# https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/d197f2717eb0d067fe35644db91c4910b6d# 46c06/analyses/fusion_filtering/06-recurrent-fusions-per-histology.R#L111
```
This link also doesn't work for me.

cansavvy commented on analyses/focal-cn-file-preparation/run-prepare-cn.sh:
```diff
@@ -44,58 +47,64 @@ Rscript --vanilla 04-prepare-cn-file.R \
-#   done
-# 
-# fi
+# Define most focal units of recurrent CNVs
```
Why is this being uncommented now?

cbethell replied:
This was done in PR #644 (it was originally only commented out for developmental purposes).

cbethell commented:
@cansavvy I am re-requesting your review because I adjusted the final output file named `results/consensus_seg_recurrent_focal_cn_units.tsv` to include the biospecimen IDs (instead of the tallied data frame without the sample IDs).

The reason behind this change is that this format would likely be more useful for downstream analyses, as I have found in the oncoprint analysis in PR #666.

The updated rendered notebook can be seen [here](https://cbethell.github.io/open-pbta-output/06-find-recurrent-calls.nb.html).

cansavvy commented on analyses/focal-cn-file-preparation/06-find-recurrent-calls.Rmd:
````diff
@@ -83,17 +83,27 @@ The `recurrence_cutoff` defined as 3 appears to be reasonable as seen in the dis
+### Filter focal CN data frame to include only recurrent observations
+
+```{r}
+# Filter the focal CN data frame to include only the recurrent observations
````
I don't think this is exactly what you want. I * think* you want to have `Kids_First_Biospecimen_ID` to be in your `group_by()` step in the first place. That way the tally's are also by sample.

cbethell replied:
If I do that, the count will be 1 for all observations. I _believe_ we want to find out which status calls occur in multiple samples for a specific region.

cansavvy replied:
Ah I see. Yes you're right. I lost track of the goal here.

cansavvy commented on analyses/focal-cn-file-preparation/06-find-recurrent-calls.Rmd:
````diff
@@ -0,0 +1,114 @@
+```{r}
+# Filter the focal CN data frame to include only the recurrent observations
+final_df_recurrent <- focal_cn_df %>%
+  select(-region_type) %>%
````
Why are we getting rid of region_type here?

cbethell replied:
I did this to avoid the duplicate column that would come from the `inner_join` step below (there's probably a better way to do this here).

cansavvy replied:
This is fine to me. Maybe just add a comment to explain why.

jaclyn-taroni replied:
I _think_ including `region_type` in your `by` below would accomplish the same thing and potentially be a little friendlier to your reader
```R
   inner_join(recurrent_df, by = c("region", "status",  "region_type"))
```

cbethell replied:
Ah, I'll give that a try @jaclyn-taroni, thank you!

cbethell commented:
@cansavvy would you mind taking one last look at this? I added the comment as you suggested in your [previous review](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/686#discussion_r425767036) and also updated the `README.md` file with a few lines regarding the notebook in this PR.

jaclyn-taroni commented:
I was looking to merge #666 and noticed `analyses/focal-cn-file-preparation/results/consensus_seg_recurrent_focal_cn_units.tsv` being added there as well as in this pull request. Can you explain the format of this file @cbethell?

cbethell commented:
> I was looking to merge #666 and noticed `analyses/focal-cn-file-preparation/results/consensus_seg_recurrent_focal_cn_units.tsv` being added there as well as in this pull request. Can you explain the format of this file @cbethell?

@jaclyn-taroni right, this file is in the same format as `analyses/focal-cn-file-preparation/results/consensus_seg_most_focal_cn_status.tsv` but including only those calls in regions that are common amongst more than three samples. Should I instead use the more general `analyses/focal-cn-file-preparation/results/consensus_seg_most_focal_cn_status.tsv` file in PR #666? There was a very slight difference in the visualization when I ran the plot using each of the two files.

jaclyn-taroni commented:
I think it's okay to filter to recurrent CNVs for the OncoPrint (you may want to control that through the gene list argument). I would expect the OncoPrint to accept something that has the format:
 
| ID | status | region |
|---|--------|-------|

And I would expect a table that keeps track of recurrence to be more like:

| region | status | count |
|-------|-------|--------|


Which you could filter to n > 3. What I am describing is similar to this file (but in this example things are counted by _histology_): https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/0985505eb18402ecac925b2038764fb2ddc2a712/analyses/fusion_filtering/results/pbta-fusion-recurrent-fusion-byhistology.tsv

You could probably incorporate histology information into that table that also includes total counts (labeled `All` below):

| region | status | short_histology | count |
|-------|--------|----------------|------|
| 7q | gain | LGAT | n |
| 7q | gain |  All | n |

cbethell commented:
> I think it's okay to filter to recurrent CNVs for the OncoPrint (you may want to control that through the gene list argument to `01-plot-oncoprint`). I would expect the OncoPrint to accept something that has the format:
> 
> ID	status	region
> And I would expect a table that keeps track of recurrence to be more like:
> 
> region	status	count
> Which you could filter to n > 3. What I am describing is similar to this file (but in this example things are counted by _histology_): https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/0985505eb18402ecac925b2038764fb2ddc2a712/analyses/fusion_filtering/results/pbta-fusion-recurrent-fusion-byhistology.tsv
> 
> You could probably incorporate histology information into that table that also includes total counts (labeled `All` below):
> 
> region	status	short_histology	count
> 7q	gain	LGAT	n
> 7q	gain	All	n

Gotcha, so I'll change the output file in this PR to not include the sample IDs but to include the histology information. But first, I will change the file used in PR #666 to be the `analyses/focal-cn-file-preparation/consensus_seg_most_focal_cn_status.tsv` file.

jaclyn-taroni commented:
> But first, I will change the file used in PR #666 to be the `analyses/focal-cn-file-preparation/consensus_seg_most_focal_cn_status.tsv` file which will then be updated with the list of recurrent regions once this PR is merged.

The first change sounds good. For the list of recurrent regions, I'd encourage you to (in an upcoming PR) do either one of these options:

* Accept a comma-delimited set of gene lists that get essentially concatenated in `01-plot-oncoprint`
* Have a step that puts all the gene lists together - this could be good as a notebook so you put lots of explanation around the lists.

I would pick an option based on what would be easier to adapt if we were to then split the OncoPrints by histology, rather than having them all in one.

jaclyn-taroni commented:
A few things I noticed when looking at the notebook locally:

* I would filter out the uncallable regions. We would expect those to be recurrent and they are not necessarily of interest here.
* Without in-depth knowledge of the steps to obtain the uncallable status calls, I might expect that uncallable counts = total number of samples. That is not always the case - is this expected based on upstream steps?

cbethell commented:
> * I would filter out the uncallable regions. We would expect those to be recurrent and they are not necessarily of interest here.

Good call, will do 👍 

> * Without in-depth knowledge of the steps to obtain the uncallable status calls, I might expect that uncallable counts = total number of samples. That is not always the case - is this expected based on upstream steps?

I'll take a deeper look into the upstream steps because the expected behavior is as you described above, uncallable counts for a region should be equal to the total number of samples because if an arm is uncallable, all samples for that arm should be uncallable. This would also mean that all samples in any cytoband on that arm should also have an uncallable status.

cbethell commented on analyses/focal-cn-file-preparation/05-define-most-focal-cn-units.Rmd:
```diff
@@ -155,7 +155,7 @@ filtered_consensus_cytoband_status <- consensus_seg_cytoband_status %>%
   # Filter the annotated CN data to include only neutral chromosome arms and disagreements
   filter(
-    dominant_arm_status %in% c("neutral", "uncallable", "unstable") |
+    dominant_arm_status %in% c("neutral", "unstable") &
```
The filtering step here was taking the all `neutral`, `unstable`, and `uncallable` arms and providing the cytoband status for these arms. Instead, we wanted to take provide only the `neutral` and `unstable` arms to this step so that we can get the cytoband status for these arms and leave the `uncallable` arms.

That being said, for each uncallable region, all samples are now labeled `uncallable`.

jashapiro replied:
As I originally understood the logic here, it was possible to have an `uncallable` arm that contained callable cytoband (i.e. an arm that was 70% uncallable could still have many callable bands), so we wanted to include those.

cbethell replied:
Ah okay, that's reasonable. I'll revert the change to the logic. Re the comment above:
> Without in-depth knowledge of the steps to obtain the uncallable status calls, I might expect that uncallable counts = total number of samples. That is not always the case - is this expected based on upstream steps?

I believe the reason the total number of samples for an uncallable region may not add up is because all samples may not have data for that region? (I am trying to track back and look at the raw data and how many samples that have data for each region)

Were the inconsistent regions the X and/or Y chromosomes? The `Xp` arm is where I noticed the inconsistency described.

jaclyn-taroni replied:
> Were the inconsistent regions the X and/or Y chromosomes? The Xp arm is where I noticed the inconsistency described.

The short answer is I don't remember - looks like you just reverted this change, is there anything you would consider an inconsistency?

cbethell replied:
So thinking about the logic a little more and @jashapiro's comment,

For sample X, cytoband 11q11 may be `uncallable` because more than 50% of the region is uncallable for that sample.

For sample Y on the other hand, cytoband 11q11 may be `loss` because more than 90% of the region for that sample is loss.

This can only be done for samples that have copy number data for that specific region (hence some regions may not have data for all 915 samples). I used the cytoband data for an example, but the same goes for the arm level calls, although arm level calls would be expected to have more agreement between samples due to the region being broader? (this is just a guess).

So based on that logic, I do not believe that the uncallable counts would necessarily be equal to the total number of samples, does that make sense?

That being said, it appears that the `Xp` and `Xq` arms are the only chromosome arms to have less samples than the total number of samples that have been labeled `uncallable`, but they are no other competing calls for these arms (by this I mean, `uncallable` is the only unique label in the `Xp` and `Xq` regions).

cbethell commented on analyses/focal-cn-file-preparation/06-find-recurrent-calls.Rmd:
```diff
@@ -0,0 +1,189 @@
+  # Filter to only gene regions
+  filter(region_type == "gene") %>%
+  # Select the top genes by count
+  top_n(20, count) %>%
```
@jaclyn-taroni if I move this step before filtering to only gene regions, it will return the top `n` for all regions. Can you further clarify what you mean here?

jaclyn-taroni replied:
The number 20 is hardcoded here, assign 20 to a variable up top.

cbethell replied:
Ah gotcha 👍
